prune/prune.go

```diff
@@ -16,184 +16,249 @@ package prune
 
-	if config.CustomStrategy == nil && config.Strategy.Mode == CustomStrategy {
-		return fmt.Errorf("custom strategies require a strategy function to be specified")
+// NewPruner returns a pruner that uses the given strategy to prune objects.
+func NewPruner(prunerClient client.Client, opts ...PrunerOption) Pruner {
+	podGVK := corev1.SchemeGroupVersion.WithKind("Pod")
+
+	jobGVK := batchv1.SchemeGroupVersion.WithKind("Job")
+
+	pruner := Pruner{
+		Registry: defaultRegistry,
+		Client:   prunerClient,
+		DryRun:   false,
+		Logger:   Logger(context.Background(), Pruner{}),
+		GVK:      podGVK,
 	}
 
-	if len(config.Namespaces) == 0 {
-		return fmt.Errorf("namespaces are required")
+	// Populate the default IsPrunableFunc(s)
+	RegisterIsPrunableFunc(podGVK, DefaultPodIsPrunable)
+
+	RegisterIsPrunableFunc(jobGVK, DefaultJobIsPrunable)
+
+	for _, opt := range opts {
+		opt(&pruner)
 	}
 
-	if containsString(config.Namespaces, "") {
-		return fmt.Errorf("empty namespace value not supported")
+	return pruner
+}
+
+// Prune runs the pruner.
+func (p Pruner) Prune(ctx context.Context) ([]client.ObjectKey, error) {
+	var objs []client.Object
+	p.Logger.Info("Starting the pruning process...")
+	listOpts := client.ListOptions{
+		LabelSelector: labels.Set(p.Labels).AsSelector(),
+		Namespace:     p.Namespace,
 	}
 
-	_, err = labels.Parse(config.LabelSelector)
-	if err != nil {
-		return err
+	var unstructuredObjs unstructured.UnstructuredList
+	unstructuredObjs.SetGroupVersionKind(p.GVK)
+	if err := p.Client.List(ctx, &unstructuredObjs, &listOpts); err != nil {
+		p.Logger.Error(err, "failed to get a list of resources for pruning", "Labels", p.Labels, "Namespace", p.Namespace)
+		return nil, fmt.Errorf("failed to get list of objects -- ERROR -- %s", err)
 	}
 
-	if config.Strategy.Mode == MaxAgeStrategy {
-		_, err = time.ParseDuration(config.Strategy.MaxAgeSetting)
+	for _, unsObj := range unstructuredObjs.Items {
+		obj, err := convert(p.Client, p.GVK, &unsObj)
 		if err != nil {
-			return err
+			return nil, err
 		}
+
+		if err := p.IsPrunable(obj); isUnprunable(err) {
+			continue
+		} else if err != nil {
+			return nil, err
+		}
+
+		objs = append(objs, obj)
+	}
+
+	objsToPrune, err := p.Strategy(ctx, objs)
+	if err != nil {
+		p.Logger.Error(err, "failed to get a list of resources to prune from Strategy")
+		return nil, fmt.Errorf("failed when running Strategy -- ERROR -- %s", err)
+	}
+
+	if p.DryRun {
+		// print out objects
+		return objsToPrune, nil
 	}
-	if config.Strategy.Mode == MaxCountStrategy {
-		if config.Strategy.MaxCountSetting < 0 {
-			return fmt.Errorf("max count is required to be greater than or equal to 0")
+
+	// Prune the resources
+	for _, obj := range objsToPrune {
+		// Prune
+		prunableObj := &unstructured.Unstructured{}
+		prunableObj.SetName(obj.Name)
+		prunableObj.SetNamespace(obj.Namespace)
+		prunableObj.SetGroupVersionKind(p.GVK)
+
+		if err = p.Client.Delete(ctx, prunableObj); err != nil {
```

Comment:
This is due to the switch I made from `type StrategyFunc func(ctx context.Context, objs []client.Object) ([]client.Object, error)` --> `type StrategyFunc func(ctx context.Context, objs []client.Object) ([]client.ObjectKey, error)`  in response to a [previous review comment you left](https://github.com/operator-framework/operator-lib/pull/105#discussion_r849668069)

However now that I think about it, I remember us speaking offline about just using the `client.Object` type instead due to us already enforcing use of controller-runtime throughout the `prune` package as well as operator-lib in general.

I think because of this it would be best to switch it back to `type StrategyFunc func(ctx context.Context, objs []client.Object) ([]client.Object, error)` so that it is easier for both the user of the `prune` package and for us to just pass the object in rather than constructing one from a `client.ObjectKey` type. 

What do you think?

Reply:
Agreed

Reply:
done in https://github.com/operator-framework/operator-lib/pull/105/commits/3f60543cf9ce9403d249d252ce3cf1bd9a42d34d